src/relay/qnn/pass/quantize_rewrite.cc

```diff
@@ -0,0 +1,262 @@
+    *shift = 0;
+    return;
+  }
+  const double q = std::frexp(double_multiplier, shift);
+  auto q_fixed = static_cast<int64_t>(std::round(q * (1ll << (idtype_bits - 1))));
+  CHECK_LE(q_fixed, (1ll << (idtype_bits - 1)));
+  if (q_fixed == (1ll << (idtype_bits - 1))) {
+    q_fixed /= 2;
+    ++*shift;
+  }
+  CHECK_LE(q_fixed, std::numeric_limits<int32_t>::max());
+  *fixed_point_multiplier = static_cast<int32_t>(q_fixed);
+}
+
+/*
+ * Requantization using only integer computation. Here, the computation is
+ * converted to a fixed point computation by computing output multiplier and
+ * shift. This is useful, if the target device does not support/have very
+ * expensive floating point computations.
+ *
+ * Original compuation is scale_fp32 * quantized_tensor.  To convert into
+ * integer computation, the multiplication with fp32 scalar can be replaced by
+ * multiplication with an int value and then right shifting the result. This
+ * approximates the floating point computation with a fixed point computation.
+ *
+ * The whole computation this can be broken down into following steps
+ * 1) Calculate the integer multiplier and integer shift.
+ * 2) Subtract the input integer point.
+ * 2) Multiply the integer fixed point multiplier with quantized tensor.
+ * 3) Round the result.
+ * 4) Right shift the result.
+ * 5) Add the output_zero_point.
+ * 6) Cast to the out_dtype.
+ *
+ */
+Expr RequantizeInt(const Expr& input_tensor,
+    const RequantizeAttrs* param, const DataType& idtype,
+    const Array<IndexExpr>& out_shape) {
+
+  double double_multiplier = param->input_scale/param->output_scale;
+
+  // The multiplication will be performed in higher precision. Find the dtype.
+  int idtype_bits = idtype.bits();
+  DataType up_idtype = Int(2 * idtype_bits);
+
+  // 1) Calculating the integer multiplier and integer shift
+  int32_t fixed_point_multiplier;
+  int shift;
+  GetFixedPointMultiplierShift(double_multiplier, &fixed_point_multiplier,
+          &shift, idtype);
+  int left_shift = shift > 0 ? shift : 0;
+  int right_shift = shift > 0 ? 0 : -shift;
+
+  // 2) Subtract the input_zero_point
+  auto tensor = input_tensor;
+  tensor = Cast(tensor, up_idtype);
+  if (param->input_zero_point != 0) {
+    auto input_zp = MakeConstantScalar(up_idtype, param->input_zero_point);
+    tensor = Subtract(tensor, input_zp);
+  }
+
+
+
+  // 3) Multiply the integer multiplier
+  if (left_shift != 0) {
+    tensor = Multiply(tensor, MakeConstantScalar(up_idtype, 1 << left_shift));
+  }
+  // Perform the multiplication in higher precision.
+  // If idtype is Int(32), the scalar is a fixed point value of int32 where the
+  // decimal point is between bits 31 and 30. After multiplying with
+  // input_tensor, the result in int64 where the decimal point is sitting
+  // between bits 31 and 30 (from the right, rightmost bit is bit 0).
+  Expr scalar = MakeConstantScalar(up_idtype, fixed_point_multiplier);
+  auto multiplied_t = Multiply(tensor, scalar);
```

Comment:
Where is `overflow` logic? If it is `overflow`, we return int32_max.

Reply:
Will look into this. I remember seeing an overflow check. Thanks for pointing out.

Reply:
@FrozenGene Thought little more about it. I think we don't need this check. Please let me know if this makes sense. 

The relevant portion from TFLite is
~~~
  bool overflow = a == b && a == std::numeric_limits<std::int32_t>::min();
~~~

Here a is input tensor and b is the fixed_point_multiplier. The fixed_point_multiplier is calculated using [frexp](https://en.cppreference.com/w/cpp/numeric/math/frexp).

Now, frexp gives negative fixed point multiplier only when the input floating point number is negative. In our case, the number is `input_scale/output_scale`. I think the scales are always positive, so we should never have negative number. (Let me know if I am wrong)

Reply:
Right.